### PR TITLE
shared vertex buffers

### PR DESCRIPTION
**Pull request by TimSylvester**
I was able to convert symbol bucket data to a `shared_ptr` and pass it through the `VertexAttribute` without too much trouble.  It ended up being more changes that I expected, but I also found and fixed some bugs along the way.

### CODE REVIEW

File: src/mbgl/gfx/vertex_vector.hpp [hunk 1]
```diff
@@ -7,16 +7,24 @@
 namespace mbgl {
 namespace gfx {
 
+struct VertexVectorBase {
```
Comment (mwilsnd) [inline src/mbgl/gfx/vertex_vector.hpp:10]: This should ether be a class type or the forward decl in `vertex_attribute.hpp` needs to change to struct

File: src/mbgl/gl/upload_pass.cpp [hunk 2]
```diff
@@ -170,21 +170,35 @@ gfx::AttributeBindingArray UploadPass::buildAttributeBindings(
...
         const auto& defaultGL = static_cast<const gl::VertexAttributeGL&>(defaultAttr);
         const auto stride = defaultAttr.getStride();
         const auto offset = static_cast<uint32_t>(allData.size());
         const auto index = static_cast<std::size_t>(defaultGL.getIndex());
 
+        bindings.resize(std::max(bindings.size(), index + 1));
+
+        if (const auto& buffer = gl::VertexAttributeGL::getSharedRawBuffer(*this, effectiveAttr, usage)) {
+            bindings[index] = {
+                /*.attribute = */ {effectiveAttr.getSharedType(), effectiveAttr.getSharedOffset()},
```
Comment (mwilsnd) [inline src/mbgl/gl/upload_pass.cpp:184]: `getSharedOffset` and `getSharedVertexOffset` should return `uint32_t` or a static_cast should be done here. (narrowing conversion warning, MSVC C4838).

**Review by mwilsnd — CHANGES_REQUESTED** [review]
Looking promising - can you rebase or merge in the latest from topic/drawable?

**Review by sjg-wdw — APPROVED** [review]
Seems like a move in the right direction for sharing those resources.

**Review by mwilsnd — COMMENTED** [review]
This is good improvement. We're down to around 40% more memory vs. the legacy renderer.

Legacy: RAM 212,576K, VRAM 207,888K
image

Drawable: RAM 414,884, VRAM 313,968K
image

Looks like there is an artifact around the edges of fill patterns:
image

**Comment by mwilsnd** [comment]
The artifact was introduced in #1288

**Review by stefankarschti — APPROVED** [review]

File: src/mbgl/gfx/drawable_builder.cpp [hunk 3]
```diff
@@ -56,25 +56,8 @@ void DrawableBuilder::flush() {
...
-                        }
-                    } else {
-                        // throw?
-                        Log::Warning(Event::General, "Invalid attribute count");
-                    }
-                }
-            });
-
+        const auto& builderAttrs = getVertexAttributes();
+        if (auto drawAttrs = builderAttrs.clone()) {
```
[the remark this answers is not in the document]
Comment (TimSylvester) [reply src/mbgl/gfx/drawable_builder.cpp:60]: Yeah, the drawables and builders could take shared references to attributes and that would be a good next step.
That's a bit problematic when the builder is setting up the attributes and might be modifying something that's being shared, but the direction we're going it's less so.

[the remark this answers is not in the document]
Comment (sjg-wdw) [reply src/mbgl/gfx/drawable_builder.cpp:60]: There are two use cases for the Builders now I think.  One is the flexible one that we envisioned originally and the other is this sort of reference and rework mode that we're using with the buckets.
We could go as far as making that explicit when setting them up.  If you're in that more strict memory mode then you don't allow the developer to add new geometry to a builder, perhaps.

**Review by mwilsnd — APPROVED** [review]

**Review by alexcristici — APPROVED** [review]

**Pull request merged**